src/DI/Container.php

```diff
@@ -163,4 +163,57 @@ public function http() {
 	public function event() {
 		return $this['ilAppEventHandler'];
 	}
+
+	/**
+	 * @return \ilIniFile
+	 */
+	public function iliasIni() {
+		return $this['ilIliasIniFile'];
+	}
+
+	/**
+	 * @return \ilIniFile
+	 */
+	public function clientIni() {
+		return $this['ilClientIniFile'];
+	}
+
+	/**
+	 *  @return \ilStyleDefinition
+	 */
+	public function systemStyle(){
+		return $this['styleDefinition'];
+	}
+
+	/**
+	 *  @return \ilHelpGUI
+	 */
+	public function help(){
+		return $this['ilHelp'];
+	}
+
+	/**
+	 * This is syntactic sugar for executing the try catch statement in the clients code.
+	 * Note that the use of the offsetSet code of the default container should be avoided,
+	 * since knowledge about the containers internal mechanism is injected.
+	 *
+	 * Example:
+	 * //This is bad since the client should not need to know about the id's name
+	 * $DIC->offsetSet("styleDefinition")
+	 *
+	 * //This is better, since the client just needs to know the name defined in the
+	 * //interface of the component
+	 * $DIC->isDependencyAvailable("systemStyle")
+	 *
+	 * @param $name
+	 * @return bool
+	 */
+	public function isDependencyAvailable($name){
```

Comment:
I see what this does, but currently don't understand why and where one would want to use this. Could you elaborate a little?

Reply:
Sure, see: https://github.com/ILIAS-eLearning/ILIAS/pull/918, where we discuss the issue a little. There are points, where a service needs to know, if another dependency is already initialized. One first step to do that is to check if it is loaded into the container. In: https://github.com/ILIAS-eLearning/ILIAS/pull/918 @chfsx proposes to use: $DIC->offsetExists("tpl"). This works, but I hate to use it, due to terrible wiring of the string "tpl" into some other component. my component should not know, that e.g. systemStyle is called "styleDefinition" in the containers internals.

Reply:
I fully agree @Amstutz that `$DIC->offsetExists("tpl")` is a bad idea, I think on the long term the best solution would be to ask if an instance für a specific interface is available. this would need that we generate interfaces for all globals we currently use which – IMHO – could be a way to discuss such an interface and after finalizing them, there is a way to write alternative instances for the same interface. Then we would need to store the instances not with names like "systemStyle" or "ilCtrl", we should store them in the container with its interfacename

Reply:
I find it super questionable that some object needs to get that information at all and strongly suspect there are some deeper issues that will be masked by that approach (and cause other problems down the line). Could we have a look at the original problem together per skype.

Reply:
@klees sure there is, as always if you start cleaning up the room at some place you find a lot more messy stuff under the carpet. however, since it's no good idea to start cleaning on all floors at once, I would rather finish one fix and then start the next. 

Such a function could work as good marker for this. However I can also live with the offset method. The tpl initialization mess should imo be tackled separately.

Reply:
IMHO the long term solution is to switch to proper dependency injection completely and stop using service locator/globals. Asking if some service exists will then not be necessary anymore.

I also think that is wrong to think of the DIC as "containing instances of interfaces" or store objects in the DIC with the name of the interface they are implementing. The DIC contains services. Services implement one or more interfaces and may depend on other services by telling which interface they need this services to implement. The DIC allows to resolve that dependencies by defining a dependency graph. The crucial inside is, that there does not need to be or even shouldn't be a one to one relation between interfaces and services.

`$ilLng` for example imo may implement multiple interfaces (e.g. one for resolving ids to text, one for setting new text for ids, ...). These could well be split, because there will be a lot of other services that only need that id-resolving capability, but not the id-definition capability. Splitting then will make it more explicit which capabalities are really needed in a service depending on some interface implementation which will in turn make unit testing easier, as well as refactorings, customizing, reimplementation. On the other hand it is well possible that multiple services implement the same interface, for `$ilLng` it would well be possible to think of the different languages as different services implementing the same set of interfaces. Also I can think of a service that only implements id-resolving but not id-definition, e.g. in a naive version that always returns the id itself as text for the id for debugging purpose.

Besides that general considerations, I'm still interested in the problem at hand. Maybe a good way forward is to introduce `isDependencyAvailable` with a big fat "do not use this in new code" or similar warning and an according dicto rule that clearly marks the usage sides as problems. To me this would be a strict requirement, because said method makes the wrong thing easy(er) which is bad in the long term.